File: 一些小demo/URLSchemeDemo/UrlSchemeDemoOC/UrlSchemeDemoOC/ViewController.m
//
//  ViewController.m
//  UrlSchemeDemoOC
//
//  Created by Changjun Li on 2024/1/2.
//

#import "ViewController.h"
#import "Masonry.h"

@interface ViewController ()

@property (nonatomic, strong) UILabel *infoLabel;
@property (nonatomic, strong) UIButton *jumpButton;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
    [self initUI];
}

- (void)onClickedJumpButton {
    NSString *urlString = [NSString stringWithFormat:@"swiftDemo://?user=elee"];
    NSURL *url = [NSURL URLWithString:urlString];
    if ([[UIApplication sharedApplication] canOpenURL:url]) {
        [[UIApplication sharedApplication] openURL:url options:@{} completionHandler:^(BOOL success) {
            if (!success) {
                self.infoLabel.text = [NSString stringWithFormat:@"跳转失败"];
            }
        }];
    } else {
        self.infoLabel.text = [NSString stringWithFormat:@"目标app不存在 or 不在白名单"];
        return;
    }
}

- (void)initUI {
    [self.view addSubview:self.infoLabel];
    [self.view addSubview:self.jumpButton];
    
    
    [_infoLabel mas_makeConstraints:^(MASConstraintMaker *make) {
        make.bottom.equalTo(_jumpButton.mas_top);
        make.height.equalTo(@50);
        make.centerX.equalTo(self.view.mas_centerX);
    }];
    [_jumpButton mas_makeConstraints:^(MASConstraintMaker *make) {
        make.centerX.equalTo(self.view.mas_centerX);
        make.centerY.equalTo(self.view.mas_centerY);
        make.width.equalTo(@150);
        make.height.equalTo(@50);
    }];
}

- (UILabel *)infoLabel {
    if (_infoLabel == nil) {
        _infoLabel = [[UILabel alloc] init];
        _infoLabel.text = @"这是OCDemo";
        _infoLabel.font = [UIFont systemFontOfSize:17];
        _infoLabel.numberOfLines = 0;
        _infoLabel.textColor = [UIColor systemGrayColor];
        _infoLabel.textAlignment = NSTextAlignmentCenter;
    }
    return  _infoLabel;
}

- (UIButton *)jumpButton {
    if (_jumpButton == nil) {
        _jumpButton = [UIButton buttonWithType:UIButtonTypeCustom];
        _jumpButton.frame = CGRectMake(100, 100, 250, 50);
        _jumpButton.layer.cornerRadius = 15;
        [_jumpButton setTitle:@"跳转" forState:UIControlStateNormal];
        [_jumpButton setTitleColor:[UIColor whiteColor] forState:UIControlStateNormal];
        [_jumpButton setBackgroundColor:[UIColor purpleColor]];
        [_jumpButton addTarget:self action:@selector(onClickedJumpButton) forControlEvents:UIControlEventTouchUpInside];
    
    }
    return _jumpButton;
}

@end

File: 框架探究demo/URLSchemeDemo/UrlSchemeDemoSwift/UrlSchemeDemoSwift/ViewController.swift
//
//  ViewController.swift
//  UrlSchemeDemoSwift
//
//  Created by Changjun Li on 2024/1/2.
//

import UIKit
import SnapKit

class ViewController: UIViewController {
    
    let infoLabel = UILabel()
    let jumpButton = UIButton()

    override func viewDidLoad() {
        super.viewDidLoad()
        // Do any additional setup after loading the view.
        initUI()
    }
    
    @objc func onClickedJumpButton() {
        let urlString = "ocDemo://?user=elee"
        guard let url = URL(string: urlString) else {
            self.infoLabel.text = "url不存在"
            return
        }
        guard UIApplication.shared.canOpenURL(url) else {
            self.infoLabel.text = "目标app不存在 or 不在白名单"
            return
        }
        UIApplication.shared.open(url) { success in
            if !success {
                self.infoLabel.text = "跳转失败"
            }
        }
    }
    
    func initUI() {
        self.view.addSubview(infoLabel)
        self.view.addSubview(jumpButton)
        
        infoLabel.text = "这是SwiftDemo"
        infoLabel.font = UIFont.systemFont(ofSize: 17)
        infoLabel.numberOfLines = 0
        infoLabel.textColor = UIColor.systemGray
        infoLabel.textAlignment = .center
        infoLabel.snp.makeConstraints { make in
            make.bottom.equalTo(jumpButton.snp.top)
            make.height.equalTo(50)
            make.centerX.equalTo(self.view.snp.centerX)
        }
        
        jumpButton.setTitle("跳转", for: .normal)
        jumpButton.layer.cornerRadius = 15
        jumpButton.tintColor = UIColor.white
        jumpButton.backgroundColor = UIColor.purple
        jumpButton.addTarget(self, action: #selector(onClickedJumpButton), for: .touchUpInside)
        jumpButton.snp.makeConstraints { make in
            make.centerX.equalTo(self.view.snp.centerX)
            make.centerY.equalTo(self.view.snp.centerY)
            make.width.equalTo(150)
            make.height.equalTo(50)
        }
    }

}

